chia-wallet/src/puzzles/nft.rs

```diff
@@ -1,5 +1,142 @@
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, ToClvm, FromClvm)]
+#[clvm(curried_args)]
+pub struct NftStateLayerArgs {
+    pub mod_hash: [u8; 32],
+    pub metadata: NodePtr,
+    pub metadata_updater_puzzle_hash: [u8; 32],
+    pub inner_puzzle: NodePtr,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, ToClvm, FromClvm)]
+#[clvm(proper_list)]
+pub struct NftStateLayerSolution {
+    pub inner_solution: NodePtr,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, ToClvm, FromClvm)]
+#[clvm(curried_args)]
+pub struct NftOwnershipLayerArgs {
+    pub mod_hash: [u8; 32],
+    pub current_owner: Option<[u8; 32]>,
+    pub transfer_program: NodePtr,
+    pub inner_puzzle: NodePtr,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, ToClvm, FromClvm)]
+#[clvm(proper_list)]
+pub struct NftOwnershipLayerSolution {
+    pub inner_solution: NodePtr,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, ToClvm, FromClvm)]
+#[clvm(curried_args)]
+pub struct NftRoyaltyTransferPuzzleArgs {
+    pub singleton_struct: SingletonStruct,
+    pub royalty_puzzle_hash: [u8; 32],
+    pub trade_price_percentage: u16,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct NftMetadata {
+    pub edition_number: u64,
+    pub edition_total: u64,
+    pub data_uris: Vec<String>,
+    pub data_hash: Option<[u8; 32]>,
+    pub metadata_uris: Vec<String>,
+    pub metadata_hash: Option<[u8; 32]>,
+    pub license_uris: Vec<String>,
+    pub license_hash: Option<[u8; 32]>,
+}
+
+impl Default for NftMetadata {
+    fn default() -> Self {
+        Self {
+            edition_number: 1,
+            edition_total: 1,
+            data_uris: Vec::new(),
+            data_hash: None,
+            metadata_uris: Vec::new(),
+            metadata_hash: None,
+            license_uris: Vec::new(),
+            license_hash: None,
+        }
+    }
+}
+
+impl FromClvm for NftMetadata {
+    fn from_clvm(a: &Allocator, node: NodePtr) -> Result<Self> {
+        let items: Vec<(String, NodePtr)> = FromClvm::from_clvm(a, node)?;
+        let mut metadata = Self::default();
+
+        for (key, value_ptr) in items {
+            match key.as_str() {
+                "sn" => metadata.edition_number = FromClvm::from_clvm(a, value_ptr)?,
+                "st" => metadata.edition_total = FromClvm::from_clvm(a, value_ptr)?,
+                "u" => metadata.data_uris = FromClvm::from_clvm(a, value_ptr)?,
+                "h" => metadata.data_hash = FromClvm::from_clvm(a, value_ptr)?,
+                "mu" => metadata.metadata_uris = FromClvm::from_clvm(a, value_ptr)?,
+                "mh" => metadata.metadata_hash = FromClvm::from_clvm(a, value_ptr)?,
+                "lu" => metadata.license_uris = FromClvm::from_clvm(a, value_ptr)?,
+                "lh" => metadata.license_hash = FromClvm::from_clvm(a, value_ptr)?,
+                _ => (),
+            }
+        }
+
+        Ok(metadata)
+    }
+}
+
+impl ToClvm for NftMetadata {
+    fn to_clvm(&self, a: &mut Allocator) -> Result<NodePtr> {
```

Comment:
I think it would be good to have round-trip tests for the types that implement custom `to_clvm()` and `from_clvm()`, to ensure they match. One think I added recently was a fuzzer for serialization round-trips.

If you implement `Arbitrary` (at least when built with fuzzing enabled) it's quite easy to make a fuzzer that turns an object into CLVM and then back again, to test round-trips.

Reply:
we only run the fuzzers for 20 seconds or so. I investigated this a while back and the CI run-time was dominated by link-time. So I disabled LTO, which sped it up quite a lot. Linking each fuzzer binary still takes quite a while though.

There's also some work to offload the fuzzing to separate machines, probably outside of github actions.

Reply:
I believe I added fuzzing correctly. I also added code to implement synthetic key derivation, so another review will be needed for these two things